scaper/core.py

```diff
@@ -358,6 +359,72 @@ def _validate_distribution(dist_tuple):
                 'number that is equal to or greater than trunc_min.')
 
 
+def _modify_source_time(source_time, source_duration, event_duration):
+    '''
+    Modify a source_time distribution tuple according to the duration of the
+    source and the duration of the event. This allows you to sample from 
+    anywhere in a source file without knowing the exact duration of every
+    source file. 
+
+    Parameters
+    ----------
+    source_time : tuple
+        Tuple specifying a distribution to sample from. See Scaper.add_event
+        for details about the expected format of the tuple and allowed values.
+    source_duration : float
+        Duration of the source audio file.
+    event_duration : float
+        Duration of the event to be extracted from the source file.
+
+    See Also
+    --------
+    Scaper.add_event : Add a foreground sound event to the foreground
+    specification.
+    '''
+    _validate_distribution(source_time)
+    old_source_time = deepcopy(source_time)
+    source_time = list(source_time)
+
+    # If it's a constant distribution, just make sure it's within bounds.
+    if source_time[0] == 'const':
+        if source_time[1] + event_duration > source_duration:
+            source_time[1] = max(0, source_duration - event_duration)
+    # If it's a choose, iterate through the list to make sure it's all in bounds.
+    elif source_time[0] == 'choose':
+        for i, t in enumerate(source_time[1]):
+            if t + event_duration > source_duration:
+                source_time[1][i] = max(0, source_duration - event_duration)
```

Comment:
Consider the following case:
* `source_duration = 5`
* `event_duration = 3`
* `source_time[1] = [0, 1, 2, 3, 4, 5]`

This current code will update the array to `[0, 1, 2, 2, 2, 2]`, meaning it's much more likely for the user to get a `source_time` of 2 as opposed to `0` or `1`. Presumably the user wants a value chosen at random from all possible values, i.e. `[0, 1, 2]`, with equal likelihood. 

To address this, should we not remove all duplicates from the modified `source_time[1]` array?

Reply:
That's a good idea, we could just make it `list(set(source_time[1]))`?

Reply:
yup

Reply:
@pseeth looks like we're still missing the conversion to `list(set(source_time[1]))` ?

Reply:
Done!

Reply:
Eek, I believe that's not the right place for this. The current location will result in the loop iterating over different list indices compared to those of source_time[1] which I think leads to erroneous behavior.

I meant we should add  `source_time[1] = list(set(source_time[1]))` *after* the loop which updates each value in the list, i.e. right after line 398 and outside the scope of the for loop (but inside the scope of the elif.